### PR TITLE
Support for setting PHP settings through configuration

### PR DESCRIPTION
**Pull request by nikolaposa**
Per-environment PHP configuration is very common use-case and so far it has been usually implemented by placing inline `ini_set` calls within some `php_settings.php` configuration itself. It would be much nicer if we can have clean configurations, for example:
```php
return [
    'php_settings' => [
        'display_startup_errors'        => false,
        'display_errors'                => false,
        'date.timezone'                 => 'Europe/Belgrade'
    ]
];
```
In order to achieve that, PHP settings must be set at some point, and I think that the most suitable context is the `ApplicationFactory`.

### CODE REVIEW

File: test/Container/ApplicationFactoryTest.php [hunk 1]
```diff
@@ -65,6 +65,60 @@ public function getRouterFromApplication(Application $app)
...
+            ->willReturn(true);
+
+        $this->container
+            ->get('config')
+            ->willReturn($config);
+
+        $this->factory->__invoke($this->container->reveal());
+
+        foreach ($config['php_settings'] as $name => $value) {
+            $this->assertEquals($value, ini_get($name));
```
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: this foreach should be wrapped in a try/finally where try is this loop and finally is the tear down loop

File: test/Container/ApplicationFactoryTest.php [hunk 2]
```diff
@@ -65,6 +65,60 @@ public function getRouterFromApplication(Application $app)
         return $r->getValue($app);
     }
 
+    public function testFactorySetsPhpSettingsFromConfig()
+    {
+        $this->container
+            ->has('Zend\Expressive\Router\RouterInterface')
+            ->willReturn(false);
+        $this->container
```
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: This expectation is unrelated with the feature under test. Only adds noise.

File: test/Container/ApplicationFactoryTest.php [hunk 3]
```diff
@@ -65,6 +65,60 @@ public function getRouterFromApplication(Application $app)
...
+            ->get('Zend\Expressive\Router\RouterInterface')
+            ->shouldNotBeCalled();
+
+        $this->container
+            ->has('Zend\Diactoros\Response\EmitterInterface')
+            ->willReturn(false);
+        $this->container
```
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: Remove this expectation.

File: test/Container/ApplicationFactoryTest.php [hunk 4]
```diff
@@ -65,6 +65,60 @@ public function getRouterFromApplication(Application $app)
...
+            ->get('Zend\Diactoros\Response\EmitterInterface')
+            ->shouldNotBeCalled();
+
+        $this->container
+            ->has('Zend\Expressive\FinalHandler')
+            ->willReturn(false);
+        $this->container
```
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: Remove this expectation.

File: test/Container/ApplicationFactoryTest.php [hunk 5]
```diff
@@ -65,6 +65,60 @@ public function getRouterFromApplication(Application $app)
         return $r->getValue($app);
     }
 
+    public function testFactorySetsPhpSettingsFromConfig()
+    {
+        $this->container
```
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: Is this really needed?

File: src/Container/ApplicationFactory.php [hunk 6]
```diff
@@ -144,6 +145,24 @@ public function __invoke(ContainerInterface $container)
...
     /**
+     * Sets provided PHP settings, if any.
+     *
+     * @param ContainerInterface $container
+     */
+    private function setPhpSettings(ContainerInterface $container)
+    {
+        $config = $container->has('config') ? $container->get('config') : [];
+
+        if (!isset($config['php_settings']) || !is_array($config['php_settings'])) {
```
Comment (Maks3w) [inline src/Container/ApplicationFactory.php]: if the value is not an array I guess an exception should be raised instead silently ignore.

File: src/Container/ApplicationFactory.php [hunk 7]
```diff
@@ -144,6 +145,24 @@ public function __invoke(ContainerInterface $container)
...
+            return;
+        }
+
+        foreach ($config['php_settings'] as $name => $value) {
+            ini_set($name, $value);
```
Comment (Maks3w) [inline src/Container/ApplicationFactory.php]: Should raise an exception if `ini_set` fails?

File: test/Container/ApplicationFactoryTest.php [hunk 8]
```diff
@@ -65,6 +65,60 @@ public function getRouterFromApplication(Application $app)
...
+            ->shouldNotBeCalled();
+
+        $config = [
+            'php_settings' => [
+                'display_errors' => 1,
+                'date.timezone' => 'Europe/Belgrade',
+            ],
+        ];
+
+        $currentPhpSettings = [];
```
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: It's better use PHPUnit `$this->iniSet()` and set a different value than the expected so:

* You can verify the ini set really has changed
* You can  remove both loops (this one and the restore one) and simplify the code.

**Comment by Maks3w** [comment]
TBH I think environment configuration is something must be do in the configuration scripts (vagrant, docker, ...)

In special since there are many settings can't be changed at runtime.

File: src/Container/ApplicationFactory.php (hunk 6 above)
Comment (nikolaposa) [inline src/Container/ApplicationFactory.php]: I kinda imitated `injectPreMiddleware()` flow in this case, but I guess it exception should be thrown.

File: src/Container/ApplicationFactory.php (hunk 7 above)
Comment (nikolaposa) [inline src/Container/ApplicationFactory.php]: Ok, I agree, exception can be raised there.

File: test/Container/ApplicationFactoryTest.php (hunk 2 above)
Comment (nikolaposa) [inline test/Container/ApplicationFactoryTest.php]: I agree that it adds noise, but for some reason I couldn't bypass those opening lines within the `ApplicationFactory::__invoke()`. I'll revisit that anyhow.

File: test/Container/ApplicationFactoryTest.php (hunk 8 above)
Comment (nikolaposa) [inline test/Container/ApplicationFactoryTest.php]: Thanks for the tip

File: test/Container/ApplicationFactoryTest.php (hunk 2 above)
Comment (nikolaposa) [inline test/Container/ApplicationFactoryTest.php]: @Maks3w Any idea how to bypass this error when I remove all those expectations:
```
Prophecy\Exception\Call\UnexpectedCallException: Method call:
  - has("Zend\Expressive\Router\RouterInterface")
on Double\ContainerInterface\P65 was not expected,
```
?
It seems like stub yields error when unexpected method call is made.

Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: This is related with the previous statement. https://github.com/zendframework/zend-expressive/pull/193/files#r45551040

Ok, it must be as you wrote.

File: src/Container/ApplicationFactory.php (hunk 6 above)
Comment (Maks3w) [inline src/Container/ApplicationFactory.php]: /cc @weierophinney I think all those type checks should throw Exception instead of silently ignore.

File: test/Container/ApplicationFactoryTest.php [hunk 9]
```diff
@@ -27,6 +27,68 @@ public function setUp()
         $this->factory   = new ApplicationFactory();
     }
 
+    protected function getContainer($config = null, $router = null, $finalHandler = null, $emitter = null)
```
Comment (nikolaposa) [inline test/Container/ApplicationFactoryTest.php]: @Maks3w What do think about this approach, of having a helper method for configuring container based on the test specificity?

File: test/Container/ApplicationFactoryTest.php [hunk 10]
```diff
@@ -65,6 +127,47 @@ public function getRouterFromApplication(Application $app)
...
+                'date.timezone' => 'Europe/Belgrade',
+            ],
+        ];
+
+        $this->factory->__invoke($this->getContainer($config)->reveal());
+
+        foreach ($config['php_settings'] as $name => $value) {
+            $this->assertEquals($value, ini_get($name));
+            ini_restore($name);
+        }
```
Comment (nikolaposa) [inline test/Container/ApplicationFactoryTest.php]: @Maks3w Is this in line with what you mentioned about having only one loop?

File: test/Container/ApplicationFactoryTest.php (hunk 9 above)
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: While I saw too a helper or many helper methods are needed I disagree with this proposal because add too much logic.

The problem with add logic branches in a test case (if, for, ...) is difficult to trust is bugfree. What test method test getContainer works as expected?

File: test/Container/ApplicationFactoryTest.php (hunk 10 above)
Comment (Maks3w) [inline test/Container/ApplicationFactoryTest.php]: PHPUnit assertions throw exceptions when assertion fail. So ini_restore won't be restored when test fail so next tests in the run will be executed with a wrong environment setting.

After $config

```php
   $this->iniSet('display_errors', 0);
   $this->iniSet('date.timezone', 'UTC');
```

File: src/Container/ApplicationFactory.php [hunk 11]
```diff
@@ -144,6 +146,30 @@ public function __invoke(ContainerInterface $container)
...
+     *
+     * @param ContainerInterface $container
+     * @throws Exception\InvalidArgumentException for invalid PHP settings.
+     * @throws PhpSettingsFailureException if setting of PHP configuration fails.
+     */
+    private function setPhpSettings(ContainerInterface $container)
+    {
+        $config = $container->has('config') ? $container->get('config') : [];
+
+        if (isset($config['php_settings'])) {
```
Comment (Maks3w) [inline src/Container/ApplicationFactory.php]: We prefer the early return approach. if not isset then return;

File: test/Container/ApplicationFactoryTest.php (hunk 9 above)
Comment (nikolaposa) [inline test/Container/ApplicationFactoryTest.php]: But this would allow us to have only one place for managing and maintaining container configuration and it also solves code duplication in all the test cases in this class. I agree though that it is somewhat unreliable due to its complexity, but after all, every test case which consumes will fail if it does not work as expected.

File: src/Container/ApplicationFactory.php [hunk 12]
```diff
@@ -136,6 +137,7 @@ public function __invoke(ContainerInterface $container)
 
         $app = new Application($router, $container, $finalHandler, $emitter);
 
+        $this->setPhpSettings($container);
```
Comment (nikolaposa) [inline src/Container/ApplicationFactory.php]: @Maks3w What do you think of moving `$this->setPhpSettings($container);` at the very begging of a `__invoke()` method, because it doesn't need `$app` at all?

**Comment by xtreamwayz** [comment]
Sure it would look nice, but needed? In stead you can easily add a ./config/autoload/local.php file with a few ini_set calls in it.

Or maybe add some logic in the config loader that detects and sets the php_settings. This way the settings are set even before the application is started. In case of ``display_startup_errors`` and ``display_errors`` this is what you probably want anyway. I mean if you want to set ``display_errors`` to ``false`` and there is an error before it is actually set to ``false``, the error is displayed.

**Comment by nikolaposa** [comment]
> file with a few ini_set calls in it.
The very intent of this PR was to avoid just that, I explained it in the introductory comment.

> This way the settings are set even before the application is started
I think that there is no point to track errors before this line: https://github.com/zendframework/zend-expressive-skeleton/blob/master/public/index.php#L18. Errors in pre-run stage can happen within the framework itself, unless of course someone put some massive, dynamic logic into configurations.

**Comment by Maks3w** [comment]
The alternative is use a different front controller (index_dev.php) with this kind of customizations.

For me the problem is this feature does not work with any setting, only settings which allow runtime changes.

**Comment by nikolaposa** [comment]
Ok, as for me, feel free to close this PR if you think that there will be no more comments or evaluations.

**Comment by nikolaposa** [comment]
... btw maybe at least we can consider fixing those inconsistencies regarding throwing Exception vs silent ignoring (https://github.com/zendframework/zend-expressive/pull/193#discussion_r45555722) and eventually that helper method in ApplicationFactory test case which solves code duplication (https://github.com/zendframework/zend-expressive/pull/193/files#diff-80ae022ff5121fde2617d4a28d0b4351R30).

**Comment by Maks3w** [comment]
I've open the issue https://github.com/zendframework/zend-expressive/issues/196 for track the inconsistencies. 

@nikolaposa Do you want to propose a PR for fix it?

**Comment by nikolaposa** [comment]
@Maks3w Why not. I'll send PR soon.

**Comment by Maks3w** [comment]
@nikolaposa  One important thing @weierophinney point me exceptions thrown must implement `Interop\Container\Exception\ContainerException`

We could reuse https://github.com/nikolaposa/zend-expressive/blob/feature/php-settings/src/Container/Exception/InvalidServiceException.php or create an specific one.

**Comment by nikolaposa** [comment]
@Maks3w Duly noted.

**Comment by weierophinney** [comment]
I honestly don't think this should be done as part of the application factory.

INI settings are an environmental concern, and should be done typically in one of the following places, in descending order of preference:

- system-wide INI file
- virtual host
- `.htaccess`
- bootstrap script

The rationale? You want the environment completely setup before you execute any of your application code, to ensure consistency of execution. As others have noted on this thread, a number of INI settings **CANNOT** be done in runtime regardless, and those that can will affect all code following.

Considering the number of deployment options available to PHP developers today, such changes can be easily accommodated as part of deployment:

- ZPK (Zend Server packaging format) allows defining INI settings.
- Chef, Puppet, and Ansible all let you control this as part of your application container configuration.

In a worst-case scenario, I'd create a script for setting any application-specific INI settings you rely on that can be set at runtime, and including that from the top of your `public/index.php` script; however, with the plethora of other options, that really should be a worst-case scenario.

Regarding the specifics that led to this request (inability to track where bad configuration was leading to invalid application state): I believe these have now been addressed in #194 and #195, and #197 will address the final lingering issues.

**Comment by nikolaposa** [comment]
@weierophinney Thanks for this detailed explanation, now I'm convinced that this proposal has no practical value.
